torch/csrc/distributed/c10d/comm.cpp

```diff
@@ -85,8 +85,16 @@ void broadcast_coalesced(
   }
 }
 
-PythonCommHook::PythonCommHook(py::object state, py::object hook)
-    : state_(std::move(state)), hook_(std::move(hook)){};
+PythonCommHook::~PythonCommHook() {
+  py::gil_scoped_acquire ag;
+  state_.dec_ref();
+  hook_.dec_ref();
+  // Explicitly set state_ and hook_ to nullptr to prevent py::object's dtor
+  // to decref on the PyObject again.
+  // See Note [Destructing py::object] in python_ivalue.h
+  state_.ptr() = nullptr;
+  hook_.ptr() = nullptr;
+}
```

Comment:
Thanks for fixing this, did you see some failures that were caused since this was missing?

Reply:
Thanks for catching this!

Realized that there is no easy way to fix this, so I reverted this change. Realized that the definition of both constructor and destructor of a TORCH_API class should only be used in libtorch_core target in build_variables.bzl, but this cpp file belongs to libtorch_python_distributed_core_sources, which is an extension of libtorch_python_sources.

See this picture:
![image](https://user-images.githubusercontent.com/8884619/96962443-a1bead80-14bb-11eb-8a1d-011475616d2d.png)

On the other hand, not sure if there can be a workaround by somehow extending the file WindowsTorchApiMacro.h (e.g., adding a new specifier like TORCH_CPP_API) @mrshenli  is this possible?

Reply:
It seems that reverting this change still couldn't fix the compiler error on Windows platform.

Reply:
Replaced TORCH_API by TORCH_PYTHON_API for CommHookInterface.